src/bloom/utils.rs

```
@@ -188,6 +208,77 @@ impl BloomFilterType {
         }
         Ok(0)
     }
+
+    /// Serializes bloomFilter to a byte array.
+    pub fn encode_bloom_filter(&self) -> Result<Vec<u8>, BloomError> {
+        match bincode::serialize(self) {
+            Ok(vec) => {
+                let mut final_vec = Vec::with_capacity(1 + vec.len());
+                final_vec.push(BLOOM_TYPE_VERSION);
+                final_vec.extend(vec);
+                Ok(final_vec)
+            }
+            Err(_) => Err(BloomError::EncodeBloomFilterFailed),
+        }
+    }
+
+    /// Deserialize a byte array to bloom filter.
+    /// We will need to handle any current or previous version and deserializing the bytes into a bloom object of the running Module's current version `BLOOM_TYPE_VERSION`.
+    pub fn decode_bloom_filter(decoded_bytes: &[u8]) -> Result<BloomFilterType, BloomError> {
+        if decoded_bytes.is_empty() {
+            return Err(BloomError::DecodeBloomFilterFailed);
+        }
+        let version = decoded_bytes[0];
+        match version {
+            1 => {
+                // always use new version to init bloomFilterType.
+                // This is to ensure that the new fields can be recognized when the object is serialized and deserialized in the future.
+                let (expansion, fp_rate, filters): (u32, f64, Vec<BloomFilter>) =
+                    match bincode::deserialize::<(u32, f64, Vec<BloomFilter>)>(&decoded_bytes[1..])
+                    {
+                        Ok(values) => {
+                            if values.0 == 0 {
+                                return Err(BloomError::NonScalingFilterFull);
```

Since an expansion with 0 is not possible and invalid, can we return this error instead? :  `BloomError::BAD_EXPANSION`

---

fixed.